### PR TITLE
Create HW5.py

### PR DESCRIPTION
**Pull request by WongPeiFong**
@rinnifox

### CODE REVIEW

**Comment by rinnifox** [comment]
Hi!

It seems that your code does not work at this point:

```
def desc_calc(smiles_list) -> pd.DataFrame:
    return molecular_descriptors.calculate_descriptors(smiles_list)
```

Did it work for you? `molecular_descriptors.py` does not contain a function` calculate_descriptors()`. Instead, you should use `getAllDescriptors()` with a proper mode, which should be different for train and test:

```
    print("Calculating descriptors for training data...")
    train_descriptors = desc_calc(train_data['SMILES'])
    print("Calculating descriptors for prediction data...")
    pred_descriptors = desc_calc(pred_data['SMILES'])
```

**Comment by WongPeiFong** [comment]
desc_calc Function is updated. @rinnifox

**Comment by rinnifox** [comment]
Unfortunately, your code is still not working for me. 
1. It raises an error for `train_descriptors = desc_calc(train_data['SMILES'], mode='train')` as you should pass all the training data there, not only SMILES. `getAllDescriptors` works with a Dataframe, not with Series. Check please how this function and its modes work - it can handle labels by itself.
2. Then there is an error for model training/evaluation, `y_pred, metrics_values = sar_model_evaluation(train_descriptors, train_data['LogP'])` as your train data does not have a column named 'LogP'. Also, I see that you've changed train and evaluation functions so that they accept training data and labels separately, but you have then to take into account that if you're passing the label separately, then your training data (first argument) should not contain the label! 

Please check your code carefully and make it work so that you could have the PASS grade for this course :)

**Comment by WongPeiFong** [comment]
@rinnifox 
 target column name 'LogP' is changed

**Comment by rinnifox** [comment]
With this new update I'm getting an error for 
`model, y_pred, metrics_values = sar_model_evaluation(train_descriptors.drop(columns=['LogP']), train_data['LogP'])`

Did this code work for you? Please, take your time and check carefully that your submission works.

I'd actually recommend you reading the task and the initial code snippet once again, and to try one more time from the beginning filling **only the places where the action is required**. For instance, you do not need these functions at all, the task does not require you to write them:

```
def test_train_split(file, test_size_value=0.25, train_size_value=None):
    descriptors, target = descriptor_target_split(file)
    x_train, x_test, y_train, y_test = train_test_split(descriptors, target, test_size=test_size_value,
                                                        train_size=train_size_value)
    train = descriptor_target_join(x_train, y_train)
    train.reset_index(inplace=True)
    train.drop('index', axis=1, inplace=True)
    test = descriptor_target_join(x_test, y_test)
    test.reset_index(inplace=True)
    test.drop('index', axis=1, inplace=True)
    return train, test

def descriptor_target_split(file):
    target = file.loc[:, file.columns == 'LogP']
    descriptors = file.loc[:, file.columns != 'LogP']
    return descriptors, target

def descriptor_target_join(descriptors, target):
    descriptors['LogP'] = target['LogP']
    file = descriptors
```

All the utility work for splitting the dataset is located in `molecular_descriptors.py`, depending on its mode. Please check this carefully, as it seems that by introducing these functions you make the code too complex, as it requires then to refactor many other things that are already pre-defined. Of course, you could write the code in your own way, but currently this approach does not seem to be successful.

**I'm not expecting you to correct this immediately, better concentrate on the quality of the submission.** Please consider this task as a final assignment for passing the course, as currently you do not have enough points for that. So to get them you have to provide **working code** that performs all the necessary calculations.

And if something is unclear, you are always welcome to ask the questions in Telegram.

**Comment by WongPeiFong** [comment]
@rinnifox i made it simple.

**Comment by rinnifox** [comment]
Ok, now seems that the code is able to run a bit further... However, there are still some errors present.

In current version of sklearn, there's no such function as `Imputer` that you're trying to import from `regression.py`.
It should be replaced with `SimpleImputer`, which has a bit different arguments (check sklearn documentation for that). 
So please, do not import it from `regression.py` but copy to your code and make the necessary corrections.

Please also note that you again have the ['LogP'] name of the target variable column, which is in fact called 'Target', and this also causes the errors. 

A general recommendation - if you're modifying something (e.g. in case in your local file the target column is named 'LogP' instead of 'Target'), then you should provide the actual data for testing your code.

**Comment by WongPeiFong** [comment]
@rinnifox updated

**Comment by rinnifox** [comment]
You've corrected the wrong place. The problem still persists. You should remove `from regression import fit_Ridge` and instead of that put `fit_Ridge` (or another regression function) function to your code (of course, taking into consideration the actual syntaxis for the functions used there) as HW says:
```
# copy the function of your choice from regression.py and the necessary imports for it -> it will perform hyperparameters tuning on selected regression model
```

If I correct your code myself, there is still another place, where the code breaks:
`cols = model.best_estimator_.named_steps['feature_selection'].get_support(indices=True)`
 A Pipeline that is inside `fit_Ridge` function looks like this: `model = Pipeline([('impute', a), ('scaling', b), ('anova', c), ('rf', clf)])` and doesn't have a step that called 'feature selection'.

Please note, that **the next attempt will be the last one**, as I believe 5 attempts is enough to complete this task. Moreover, I **asked twice to check and test the code before the submission**, as providing a non-working code is a bad practice. However, if the code is somehow running on your local environment - then you have to upload to your repository all the files that make you able running these scripts. 

Unfortunately, I cannot give your 6.5 missing points unless your code is running without errors. Please take your **last attempt before June 27th 12.00 (gmt +3).** In case the code is still non-working, you will need to take an extra session for this course.

**Comment by WongPeiFong** [comment]
choosed another regression function (fit_Lasso) and rewritten all. @rinnifox

**Comment by rinnifox** [comment]
Accepted.
